libraries/SITL/SIM_GPS.cpp

```diff
@@ -998,198 +996,447 @@ uint32_t GPS_NOVA::CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_
-        uint16_t time_week;
-        uint8_t num_sats;
-        // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%201
-        uint8_t pos_flags_1;
-        // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%202
-        uint8_t pos_flags_2;
-        uint8_t initialized_num;
-    } pos_time {
-        GSOF_POS_TIME_TYPE,
-        GSOF_POS_TIME_LEN,
-        htobe32(gps_tow.ms),
-        htobe16(gps_tow.week),
-        d->have_lock ? _sitl->gps_numsats[instance] : uint8_t(3),
-        pos_flags_1,
-        pos_flags_2,
-        bootcount
-    };
-    static_assert(sizeof(gsof_pos_time) - (sizeof(gsof_pos_time::OUTPUT_RECORD_TYPE) + sizeof(gsof_pos_time::RECORD_LEN)) == GSOF_POS_TIME_LEN);
-
-    constexpr uint8_t GSOF_POS_TYPE = 0x02;
-    constexpr uint8_t GSOF_POS_LEN = 0x18;
-
-    const struct PACKED gsof_pos {
-        const uint8_t OUTPUT_RECORD_TYPE;
-        const uint8_t RECORD_LEN;
-        uint64_t lat;
-        uint64_t lng;
-        uint64_t alt;
-    } pos {
-        GSOF_POS_TYPE,
-        GSOF_POS_LEN,
-        gsof_pack_double(d->latitude * DEG_TO_RAD_DOUBLE),
-        gsof_pack_double(d->longitude * DEG_TO_RAD_DOUBLE),
-        gsof_pack_double(static_cast<double>(d->altitude))
-    };
-    static_assert(sizeof(gsof_pos) - (sizeof(gsof_pos::OUTPUT_RECORD_TYPE) + sizeof(gsof_pos::RECORD_LEN)) == GSOF_POS_LEN); 
-    
-    // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Velocity.html
-    constexpr uint8_t GSOF_VEL_TYPE = 0x08;
-     // use the smaller packet by ignoring local coordinate system
-    constexpr uint8_t GSOF_VEL_LEN = 0x0D;
-
-    // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Velocity%20flags
-    enum class VEL_FIELDS : uint8_t {
-        VALID = 1U << 0,
-        CONSECUTIVE_MEASUREMENTS = 1U << 1,
-        HEADING_VALID = 1U << 2,
-        RESERVED_3 = 1U << 3,
-        RESERVED_4 = 1U << 4,
-        RESERVED_5 = 1U << 5,
-        RESERVED_6 = 1U << 6,
-        RESERVED_7 = 1U << 7,
-    };
-    uint8_t vel_flags {0};   
-    if(d->have_lock) {
-        vel_flags |= uint8_t(VEL_FIELDS::VALID);
-        vel_flags |= uint8_t(VEL_FIELDS::CONSECUTIVE_MEASUREMENTS);
-        vel_flags |= uint8_t(VEL_FIELDS::HEADING_VALID);
+    if (channel_rates[uint8_t(Gsof_Msg_Record_Type::VELOCITY_DATA)] != Output_Rate::OFF){
+        const auto last_time = last_publish_ms[uint8_t(Gsof_Msg_Record_Type::VELOCITY_DATA)];
+        const auto desired_rate = channel_rates[uint8_t(Gsof_Msg_Record_Type::VELOCITY_DATA)];
+
+        if (now - last_time > RateToPeriodMs(desired_rate)) {
+            // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Velocity.html
+            // use the smaller packet by ignoring local coordinate system
+            constexpr uint8_t GSOF_VEL_LEN = 0x0D;
+
+            // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Velocity%20flags
+            enum class VEL_FIELDS : uint8_t {
```

Comment:
nitpick: we normally use camel case for enum classes I think.  All caps are used for enum elements and definitions.

Reply:
https://ardupilot.org/dev/docs/style-guide.html#enums

Yea, good point, but it's PascalCase. This was already existing, it's just indented because it's now in conditional logic. Do you still want me to change style?